File: packages/node/iii/src/context.ts
import { AsyncLocalStorage } from 'async_hooks'
import { Logger } from './logger'

export type Context = {
  logger: Logger
}

const globalStorage = new AsyncLocalStorage<Context>()

export const withContext = async <T>(fn: (context: Context) => Promise<T>, context: Context): Promise<T> => {
  return globalStorage.run(context, async () => await fn(context))
}

export const getContext = (): Context => {
  const store = globalStorage.getStore()
  if (store) {
    return store
  }

  const logger = new Logger()

  return { logger }
}

File: packages/rust/iii/src/context.rs
use crate::logger::Logger;

#[derive(Clone)]
pub struct Context {
    pub logger: Logger,
}

tokio::task_local! {
    static CONTEXT: Context;
}

pub async fn with_context<F, Fut, T>(context: Context, f: F) -> T
where
    F: FnOnce() -> Fut,
    Fut: std::future::Future<Output = T>,
{
    CONTEXT.scope(context, f()).await
}

pub fn get_context() -> Context {
    CONTEXT
        .try_with(|ctx| ctx.clone())
        .unwrap_or_else(|_| Context {
            logger: Logger::default(),
        })
}
